metricproviders/webmetric/webmetric.go

```diff
@@ -0,0 +1,203 @@
+package webmetric
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"k8s.io/client-go/util/jsonpath"
+
+	"github.com/argoproj/argo-rollouts/utils/evaluate"
+	metricutil "github.com/argoproj/argo-rollouts/utils/metric"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	//ProviderType indicates the provider is prometheus
+	ProviderType = "WebMetric"
+)
+
+// Provider contains all the required components to run a WebMetric query
+// Implements the Provider Interface
+type Provider struct {
+	logCtx     log.Entry
+	client     *http.Client
+	jsonParser *jsonpath.JSONPath
+}
+
+// Type incidates provider is a WebMetric provider
+func (p *Provider) Type() string {
+	return ProviderType
+}
+
+func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alpha1.Measurement {
+	var (
+		err error
+	)
+
+	startTime := metav1.Now()
+
+	// Measurement to pass back
+	measurement := v1alpha1.Measurement{
+		StartedAt: &startTime,
+	}
+
+	// Create request
+	request := &http.Request{
+		Method: "GET",
+	}
+
+	request.URL, err = url.Parse(metric.Provider.WebMetric.Url)
+	if err != nil {
+		return metricutil.MarkMeasurementError(measurement, err)
+	}
+
+	for _, header := range metric.Provider.WebMetric.Headers {
+		request.Header.Set(header.Key, header.Value)
+	}
+
+	// Send Request
+	response, err := p.client.Do(request)
+	if err != nil || response.StatusCode != http.StatusOK {
+		return metricutil.MarkMeasurementError(measurement, err)
+	}
+
+	value, status, err := p.parseResponse(metric, response)
+	if err != nil || response.StatusCode != http.StatusOK {
+		return metricutil.MarkMeasurementError(measurement, err)
+	}
+
+	measurement.Value = value
+	measurement.Phase = status
+	finishedTime := metav1.Now()
+	measurement.FinishedAt = &finishedTime
+
+	return measurement
+}
+
+func (p *Provider) parseResponse(metric v1alpha1.Metric, response *http.Response) (string, v1alpha1.AnalysisPhase, error) {
+	var data interface{}
+
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return "", v1alpha1.AnalysisPhaseError, fmt.Errorf("Received no bytes in response: %v", err)
+	}
+
+	err = json.Unmarshal(bodyBytes, &data)
+	if err != nil {
+		return "", v1alpha1.AnalysisPhaseError, fmt.Errorf("Could not parse JSON body: %v", err)
+	}
+
+	buf := new(bytes.Buffer)
+	err = p.jsonParser.Execute(buf, data)
+	if err != nil {
+		return "", v1alpha1.AnalysisPhaseError, fmt.Errorf("Could not find JSONPath in body: %s", err)
+	}
+	out := buf.String()
+	outAsFloat, err := strconv.ParseFloat(buf.String(), 64)
```

Comment:
I'm concerned that only accepting float64 as a result does not provide enough flexibility for the web metrics. For example, if the endpoint pinged returned a struct like ```{
      "status": "success"
}```, the web provider couldn't check if the status field is equal to success.  To achieve both goals, the jsonParser should parse the response and pass that parsed object into the evaluate condition method. That should achieve both outcomes.

Reply:
Yeah I had this in mind as well... I'll dig into the evaluation code a bit more to get a more elegant solution here.